src/dual.jl

```diff
@@ -796,6 +796,84 @@ function SpecialFunctions.gamma_inc(a::Real, d::Dual{T,<:Real}, ind::Integer) wh
+function _map_dual_components!(fvalue!, fpartial!, y::AbstractArray{DT}, x::AbstractArray{DT}) where DT <: Dual{<:Any, T} where T
+    N = npartials(DT)
+    tx = similar(x, T) 
+    ty = similar(y, T) # temporary Array{T} for fvalue!/fpartial! application
+    # y allows res to be accessed as Array{T}
+    yarr = reinterpret(reshape, T, y)
+    @assert size(yarr) == (N + 1, size(y)...)
+    ystride = size(yarr, 1)
+
+    # calculate res values
+    @inbounds for (j, v) in enumerate(x)
+        tx[j] = value(v)
+    end
+    fvalue!(ty, tx)
+    k = 1
+    @inbounds for tt in ty
+        yarr[k] = tt
+        k += ystride
+    end
+
+    # calculate each res partial
+    for i in 1:N
+        @inbounds for (j, v) in enumerate(x)
+            tx[j] = partials(v, i)
+        end
+        fpartial!(ty, tx, i)
+        k = i + 1
+        @inbounds for tt in ty
+            yarr[k] = tt
+            k += ystride
+        end
+    end
+
+    return y
+end
+
+LinearAlgebra.ldiv!(y::StridedVector{T},
+                    m::Union{LowerTriangular{<:LinearAlgebra.BlasFloat},
+                             UpperTriangular{<:LinearAlgebra.BlasFloat}},
+                    x::StridedVector{T}) where T <: Dual =
+    (ldiv!(reinterpret(reshape, valtype(T), y)', m, reinterpret(reshape, valtype(T), x)'); y)
+
+Base.:\(m::Union{LowerTriangular{<:LinearAlgebra.BlasFloat},
+                 UpperTriangular{<:LinearAlgebra.BlasFloat}},
+        x::StridedVector{<:Dual}) = ldiv!(similar(x), m, x)
+
+for MT in (StridedMatrix{<:LinearAlgebra.BlasFloat},
+           LowerTriangular{<:LinearAlgebra.BlasFloat},
+           UpperTriangular{<:LinearAlgebra.BlasFloat},
+           SparseMatrixCSC{<:LinearAlgebra.BlasFloat}
+           )
+
+    @eval LinearAlgebra.ldiv!(y::StridedMatrix{T}, m::$MT, x::StridedMatrix{T}) where T <: Dual =
+        _map_dual_components!((y, x) -> ldiv!(y, m, x), (y, x, _) -> ldiv!(y, m, x), y, x)
+
+    @eval Base.:\(m::$MT, x::StridedMatrix{<:Dual}) = ldiv!(similar(x), m, x)
+
+    @eval LinearAlgebra.mul!(C::StridedVector{T}, A::$MT, B::StridedVector{T}) where T <: Dual =
+        (mul!(reinterpret(reshape, valtype(T), C), reinterpret(reshape, valtype(T), B), A'); C)
+
+    @eval LinearAlgebra.mul!(C::StridedVector{T}, A::$MT, B::StridedVector{T},
+                             α::Union{LinearAlgebra.BlasFloat, Integer},
+                             β::Union{LinearAlgebra.BlasFloat, Integer}) where T <: Dual =
+        (mul!(reinterpret(reshape, valtype(T), C), reinterpret(reshape, valtype(T), B), A', α, β); C)
+
+    @eval Base.:*(m::$MT, x::StridedVector{<:Dual}) = mul!(similar(x, (size(m, 1),)), m, x)
+
+    @eval LinearAlgebra.mul!(y::StridedMatrix{T}, m::$MT, x::StridedMatrix{T}) where T <: Dual =
+        _map_dual_components!((y, x) -> mul!(y, m, x), (y, x, _) -> mul!(y, m, x), y, x)
+
+    @eval Base.:*(m::$MT, x::StridedMatrix{<:Dual}) = mul!(similar(x, (size(m, 1), size(x, 2))), m, x)
```

Comment:
@mcabbott Unfortunately, I also start to loose track, but AFAIR I was exploring these possibilities:
  * `reinterpret()` would only work for the dual vectors (translating it into normal matrix multiplication), but for dual matrices there's no representation as a normal matrix linalg operation
  * `*`  and `\` methods are required to overload Base Julia methods that would promote eltypes of all vectors and matrices to *Dual*, so we need to intercept that code path early
  * `mul!()`: AFAIR there are some 3-arg methods that don't call 5-arg methods, plus it was evolving from Julia 1.0 to 1.8, so it is hard to come up with the set of methods that are optimal for all the versions. This is a part of the PR that could be potentially more polished, but as the whole infrastructure of `mul!()/ldiv!()` methods in *LinearAlgrebra* is nontrivial, I was waiting for the *ForwardDiff* devs feedback and approval of the PR in principle before going forward.

Reply:
Eventually we'll get there!

I do think you can reinterpret dual-mat * matorvec. `reinterpret(Float64, rand(3,4).+im) * rand(4, 5)` makes a  6×5 Matrix{Float64} via existing methods, I think you're suggesting that there is no method which makes a 2×3×5 array, but this is just reshape. I think that `@less mul!((rand(3).+im), (rand(3,3).+im), rand(3), true, false)` works like this, without ever calling `reinterpret` but just passing pointers. 

Agree that some of the other methods here need to catch * or / directly. 

Re 1.0, the only reason not to drop it is not being sure what version of ForwardDiff to call that... I mean it shouldn't break on 1.0, but it's totally fine for 1.0 to miss these fast paths.

Reply:
@mcabbott BLAS has complex matrix-matrix multiplication ([*zgemm()*](https://github.com/JuliaLang/julia/blob/master/stdlib/LinearAlgebra/src/blas.jl#L1495)), so there's need for reinterpretation. The *K×L* matrix *z* of duals with *N* partials could be represented as *(N+1)×L×M* array *Z*. When multiplied by the constant *K×L* matrix *A*, the result *y = A×z* could be reinterpreted as *(N+1)×K×M* array. My point is that *K* and *L* are the middle dimension of *Y* and *Z* arrays, resp. So any combination of reshaping and transposing operations would make these dimensions the first or the last one to (so that the resulting matrix is compatible for matrix *A* multiplication). One would need *permutedims*, which involves elements reshuffling.